File: Sources/BlitzBi/BlitzBiEventRepository.m
//
//  BlitzBiEventRepository.m
//  BlitzBi
//
//  Created by Akash on 01/10/20.
//  Copyright © 2016 Blitz, Inc. All rights reserved.
//

#import "BlitzBiEventRepository.h"
#import "BlitzRequestBuilder.h"

@interface BlitzBiEventRepository()
- (NSMutableDictionary*) getRequestHeaders;
@end

@implementation BlitzBiEventRepository
- (instancetype)init:(NSString*)appId
                    :(NSString*)appToken
                    :(id <PBlitzDataTransferService>)networkService {
    if (self = [super init]) {
        self->appId = appId;
        self->appToken = appToken;
        self->networkService = networkService;
    }
    return self;
}

- (void)processJsonRequest:(NSString*)url
                          :(NSData*)data
                          :(void(^)(NSObject *, NSError *))completion{
    BlitzRequestBuilder *requestBuilder = [[BlitzRequestBuilder alloc] init];
    [requestBuilder setMethod:BLITZ_POST_METHOD];
    [requestBuilder setBaseUrl:url];
    [requestBuilder setPath:[@"app/{appId}/events" stringByReplacingOccurrencesOfString:@"{appId}" withString:appId]];
    [requestBuilder setHeaders:[self getRequestHeaders]];
    [requestBuilder setHttpBody:data];
    [requestBuilder setReqType:BI_REQUEST];
    [requestBuilder setContentType:@"application/json"];
    
    [networkService executeServerCall:requestBuilder withCompletion: completion];
}

- (void)processJsonRequestWithoutResponse:(NSString*)url
                                         :(NSData*)data
                                         :(BOOL)isEmergency {
    BlitzRequestBuilder *requestBuilder = [[BlitzRequestBuilder alloc] init];
    [requestBuilder setMethod:BLITZ_POST_METHOD];
    [requestBuilder setBaseUrl:url];
    [requestBuilder setPath:[@"app/{appId}/events" stringByReplacingOccurrencesOfString:@"{appId}" withString:appId]];
    [requestBuilder setHeaders:[self getRequestHeaders]];
    [requestBuilder setHttpBody:data];
    [requestBuilder setReqType:BI_REQUEST];
    [requestBuilder setContentType:@"application/json"];
    
    if (isEmergency && ![NSThread isMainThread]) {
        [[[NSURLSession sharedSession] dataTaskWithRequest:[requestBuilder generateRequest] completionHandler:^(__attribute__((unused)) NSData *data,__attribute__((unused)) NSURLResponse *response, __attribute__((unused)) NSError *error){
            
        }] resume];
    } else {
        [networkService executeServerCallWithNoCallBack:requestBuilder];
    }
}

- (NSMutableDictionary*) getRequestHeaders {
    NSMutableDictionary *headers = [[NSMutableDictionary alloc] init];
    [headers setValue:appId forKey:@"blitzAppId"];
    [headers setValue:appToken forKey:@"blitzAppToken"];
    return headers;
}
@end

File: BI/Module/BlitzBIEventRepository.swift
//
//  BIEventRepository.swift
//  Bi
//
//  Created by Saurabh on 16/09/20.
//

import Foundation

protocol IBlitzBIEventRepository {
    func processJsonRequestWithoutResponse(url:String, body:Data?, isEmergency:Bool)
    func processJsonRequest(url:String, data:Data?, completionHandler:@escaping(BiResponse?,Error?) -> Void)
}

class BlitzBIEventRepository:IBlitzBIEventRepository {
    
    private var networkService:PBlitzDataTransferService
    
    init(networkService:PBlitzDataTransferService) {
        self.networkService = networkService
    }
    
    func processJsonRequest(url:String, data:Data?, completionHandler:@escaping(BiResponse?,Error?) -> Void) {
          let requestBuilder = BlitzRequestBuilder()
          requestBuilder.method = "POST"
          let path = ""
          requestBuilder.baseUrl = url
          requestBuilder.path = path
          requestBuilder.contentType = "application/json"
          requestBuilder.httpBody = data
          requestBuilder.reqType = BI_REQUEST
          networkService.executeServerCall(requestBuilder: requestBuilder, type: BiResponse.self, completionBlock: {result in
            switch(result) {
            case .success(let result):
                completionHandler(result,nil)
            case .failure(let error):
                completionHandler(nil,error)
            }
          })
      }
    
    func processJsonRequestWithoutResponse(url:String, body:Data?, isEmergency:Bool) {
        let requestBuilder = BlitzRequestBuilder()
        requestBuilder.method = "POST"
        let path = ""
        requestBuilder.baseUrl = url
        requestBuilder.path = path
        requestBuilder.contentType = "application/json"
        requestBuilder.httpBody = body
        requestBuilder.reqType = BI_REQUEST
        requestBuilder.responseListener = nil
        if (isEmergency  && !Thread.isMainThread) {
            do {
                try NSURLConnection.sendSynchronousRequest(requestBuilder.generateRequest() as URLRequest, returning: nil)
            } catch {
            }
        } else {
            networkService.executeServerCallWithNoCallBack(requestBuilder:requestBuilder)
        }
    }
}
